### ISSUE TITLE
Tracking Issue for float_minimum_maximum

### ISSUE THREAD

**Issue by Urgau** [rust-lang/rust#91079]
Feature gate: `#![feature(float_minimum_maximum)]`

This is a tracking issue for `f{32,64}::{min,max}imum` functions.

It provide IEEE 754-2019 `minimun` and `maximum` functions for f32/f64.

### Public API

```rust
impl f32 {
    pub fn maximum(self, other: f32) -> f32;
    pub fn minimum(self, other: f32) -> f32;
}

impl f64 {
    pub fn maximum(self, other: f64) -> f64;
    pub fn minimum(self, other: f64) -> f64;
}
```

### Steps / History

- [ ] Implementation: https://github.com/rust-lang/rust/pull/91008
- [ ] Final comment period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

**Comment by CryZe** [comment]
Just so it's tracked here: This currently does not map to the corresponding instruction on WASM and I believe Aarch64 has an instruction for this too, which it also doesn't map to.

**Comment by workingjubilee** [comment]
I believe there is an LLVM instruction for this. We can teach Rust to emit the appropriate LLVMIR directly, at the cost of adding the weight of an intrinsic.

**Comment by scottmcm** [comment]
LLVM does had an intrinsic for this, but as the PR that added these showed (<https://godbolt.org/z/czz71M9KT>), it doesn't actually work yet, which is why rustc isn't using it to implement these.

**Comment by workingjubilee** [comment]
Ah, Damn.

**Comment by rben01** [comment]
*Everything that follows will be stated in terms of the `min` and `minimum` functions but applies equally to `max` and `maximum`.*

Does the function have to be called `minimum`? I know that's the IEEE name but I think I'd find having both `min` and `minimum` confusing, as their names don't really say which is which. "The function that requires both inputs be `NaN` for its output to be `NaN` is the one with the shorter name" is not a natural rule to remember IMO. I think I'd prefer having `minimum` be called (say) `partial_min` to suggest that it follow the same rules as `PartialEq` and `PartialOrd`, which is that a single `NaN` in the input is sufficient to produce a missing output.

**Comment by m-ou-se** [comment]
Nominated at the request of @Urgau.

**Comment by m-ou-se** [comment]
The confusion between `maximum` vs `max` is definitely a problem, so we'll need to do something about that before we can stabilize that. The `max` and `min` functions are already stable, so we have limited flexibility.

We'll need some more input: How do other languages and libraries handle this? How are these named in the standard? What are the use cases for `max` vs `maximum`?

**Comment by alistaircarscadden** [comment]
I agree that having pairs of functions such that `min != minimum` and `max != maximum` is an issue. I had to check the documentation for why there were two functions to do the same thing to realize that they were functionally different. I think whatever the most explicit possible name is should be chosen. I don't think `minimum / maximum` is correct.

Here are some suggestions to add to the brainstorm. Any of these can be `min` or `minimum` and apply to both `min` and `max`.

* `min_nan`,
* `min_propigate_nans`
* `min_any_nan`
* `min_ieee_754_2019`
* `min_nans_win`
* etc.

I think that `min` with some suffix is preferable to `minimum` because it shows it is an alteration of the existing `min` function, not a competing function. I don't know the whole history of relevance of IEEE 754 2019 but I think choosing `minimum` because that's how it is named in that standard is not clear to the average Rust programmer.

**Comment by rben01** [comment]
To throw some other ideas in the mix:

By analogy with the distinction between quiet NaN and signaling NaN, maybe the new minimum (maximum) function should be called `signaling_min` (`signaling_max`). 

I also like `lesser_of` (`greater_of`), as it kind of implies that `<` (`>`) will be used, which obviously return NaN if at least one argument is NaN. You could imagine the following conversation: "Which is the lesser of 1 and NaN?" "That makes no sense, they can't be compared." "Exactly."

IMO `min_all`/`min_both` (`max_all`/`max_both`) also get the idea across -- *all*/*both* of the arguments will affect the output. In particular, even if one of the args is -inf (+inf) you still have to look at the other.

**Comment by CryZe** [comment]
This might also influence the naming situation: There's more than just these two algorithms.

Rust's max is IEEE Std 754-2008 maxNum
This new one is IEEE Std 754-2019 maximum
And there's also IEEE Std 754-2019 maximumNumber
And I feel like WebAssembly's pmax is also worth considering having (ignore's NaN for a fast impl)

**Comment by scottmcm** [comment]
The new one is arguably the "best" one, as it's the one that behaves most similarly to everything else with respect to NANs -- the other binary functions like `add` and `atan2` do the same "NAN if _either_ input is NAN", same as what `maximum` does.

So one option might be to deprecate `max` under that name and offer it as `max_ignoring_nans` or something, letting the new one get the name `maximum`.

(Thought that does have the obvious downside of being different from the `max` method on everything that's `Ord`.)

**Comment by Urgau** [comment]
> How do other languages and libraries handle this?

C/C++/Swift doesn't (yet ?). C#/Python have the IEEE 754-2019 `minimum` and `maximum` semantics as their `min`/`max` function from the start. BUT there doesn't seems to have any language that have both the 2008 and 2019 semantics.

> How are these named in the standard?

They are named `minimum` and `maximum` in IEEE 754-2019.
Note that the previous functions `minNum` and `maxNum` from IEEE 754-2008 (and below) are deprecated in IEEE 754-2019 in favor of the new one.

> What are the use cases for `max` vs `maximum`?

*(I don't really understand the question)* but I would say none. Only one should exist, in IEEE 754-2019 they deprecated `minNum` and `maxNum` in favor of `minimum` and `maximum, maybe we should do the same.

> So one option might be to deprecate `max` under that name and offer it as `max_ignoring_nans` or something, letting the new one get the name `maximum`.

+1

---

Let me know if you need more input. Happy to help.

**Comment by CryZe** [comment]
> Only one should exist, in IEEE 754-2019 they deprecated minNum and maxNum in favor of minimum and `maximum, maybe we should do the same.

I still feel like the standard is being misrepresented here. There's 4 functions now, not 2:

![https://i.imgur.com/Nl3tXju.png](https://i.imgur.com/Nl3tXju.png)

**Comment by scottmcm** [comment]
@m-ou-se 

> What are the use cases for `max` vs `maximum`?

When used in isolation as a normal binary function between two calculated values, I'd [say](https://github.com/rust-lang/rust/issues/91079#issuecomment-1226899050) that `maximum` is the one that does what people want and expect.

But when considered as a way to collapse a full sequence, both are usable in different cases.  You can think of them kinda like `.flatten().fold()` vs `.try_fold()` on `Option`s:
- `.fold(f32::NAN, f32::max)` will give you the largest non-NAN value if there is one, otherwise you get NAN (either because they're all NAN or the iterator is empty)
- `.reduce(f32::maximum)` will give you the largest value if there are *no* NANs in the iterator, `Some(NAN)` if there's one or more NANs, and `None` if the iterator is empty.

**Comment by m-ou-se** [comment]
Thanks for the clarifications. :)

The names `maximum` and `maximumNumber` from the standard are also a bit confusing, although at least the "number" part clarifies a bit that it tries its best to return a number instead of NaN. ^^'

> So one option might be to deprecate max under that name

I agree it's a shame that we already have our current `max` method doing what it does. I'm not entirely sure we should deprecate such a common method, but it is a tempting way to make space for a better solution..

> Also here's a list I made quite a while ago that lists the different instructions and languages: https://gist.github.com/CryZe/30cc76f4629cb0846d5a9b8d13144649

Thanks! It's a great overview, but the data is somewhat depressing. :(

**Comment by jsimmons** [comment]
I'm not sure it's a good idea to deprecate max/min with their existing names given that the segment of code which is likely to care about the difference in semantics between 2008 and 2019 is rather small. It's not exactly an upgrade for people using the API, just different. So lots of downstream churn for no advantage.

Seems like it would be better just to flip to the new semantics with an edition bump, if that's possible. Though it will likely break anyone who was relying on the performance of min and max being acceptable on arm [compiler explorer](https://rust.godbolt.org/z/PEYn8a49E). (Won't matter for x86 because you're already forced to write your own min and max to get good codegen there)

**Comment by CryZe** [comment]
> Though it will likely break anyone who was relying on the performance of min and max being acceptable on arm

That's "not true", as ARM has instructions for both 2019 behaviors as seen here: https://gist.github.com/CryZe/30cc76f4629cb0846d5a9b8d13144649

What's interesting to me though is that LLVM compiles the current behavior to `FMINNM / FMAXNM`, even though that's not matching the 2008 standard. The ARM documentation states that these instructions behave differently based on the value in the `AH` register. I'm currently looking into what exactly the changes are.

**Comment by jsimmons** [comment]
I'm not sure that gist is really correct. The arm docs for FMINNM state.

"NaNs are handled according to the IEEE 754-2008 standard. If one vector element is numeric and the other is a quiet NaN, the result that is placed in the vector is the numerical value, otherwise the result is identical to FMIN (scalar)."

-------- Original Message --------

**Comment by CryZe** [comment]
That's correct. The difference in behavior is in the handling of -0 and +0. That's where the alternative floating point extension comes (switches from 2019 mode to 2008) in. The difference in NaN behavior is supported by ARM as `FMAX` and `FMAXNM`.

**Comment by CryZe** [comment]
I ran it on both qemu and actual hardware to confirm that what I said above is indeed true and FMAXNM is indeed following the 2019 semantics and LLVM emits the wrong instruction:

![https://i.imgur.com/TxCndLb.png](https://i.imgur.com/TxCndLb.png)

```rust
#[inline(never)]
fn hardware_max(a: f32, b: f32) -> f32 {
    a.max(b)
}

#[inline(always)]
fn llvm_evaluated_max(a: f32, b: f32) -> f32 {
    a.max(b)
}

fn main() {
    dbg!(hardware_max(0.0, 0.0).is_sign_positive());
    dbg!(hardware_max(0.0, -0.0).is_sign_positive());
    dbg!(hardware_max(-0.0, 0.0).is_sign_positive());
    dbg!(hardware_max(-0.0, -0.0).is_sign_positive());
    println!();
    dbg!(llvm_evaluated_max(0.0, 0.0).is_sign_positive());
    dbg!(llvm_evaluated_max(0.0, -0.0).is_sign_positive());
    dbg!(llvm_evaluated_max(-0.0, 0.0).is_sign_positive());
    dbg!(llvm_evaluated_max(-0.0, -0.0).is_sign_positive());
}
```

**Comment by jsimmons** [comment]
It does not follow the 2019 semantics for NaN though? Or am I mistaken about what 2019 says. e.g it should be returning NaN if either argument is NaN. As opposed to 2008 semantics where it returns the *other* value if either argument is NaN.

-------- Original Message --------

**Comment by CryZe** [comment]
The 2019 standard has both NaN absorbtion (`maximumNumber`) and NaN propagation (`maximum`). I'm saying that Rust's `f32::max` maps to `FMAXNM` on ARM, which does the 2019 `maximumNumber` instead of the 2008 `maxNum`.

**Comment by nikic** [comment]
@CryZe I think your comments are causing unnecessary confusion here. IEEE-2008 leaves the behavior for signed zero unspecified (either operand may be returned). An instruction that follows IEEE-2019 signed zero behavior also conforms to IEEE-2008 -- just not the other way around. LLVM is correct to use FMAXNM to lower `llvm.maxnum`. There is no "wrong instruction" being used.

**Comment by CryZe** [comment]
Yeah you are right I was under the impression that it actually specified it to choose either the first or second operand, but it seems like it's implementation defined in the 2008 standard, the C standard, technically in Rust and in LLVM. I adjusted my list to reflect that.

**Comment by programmerjake** [comment]
> So one option would be to upgrade the `f32/f64::min/max` to the 2019 behavior as that's a backwards compatible change (though LLVM doesn't have an intrinsic for it atm).

it isn't actually backward compatible because IEEE 754-2008 `maxNum` behaves differently than IEEE 754-2019 `maximumNumber` for signalling NaNs: `maxNum` is defined to return a quiet NaN if either input is a signalling NaN, whereas `maximumNumber` returns non-NaN if either input is non-NaN, even if the other input is a signalling NaN.

background on min/max in IEEE 754-2019: https://grouper.ieee.org/groups/msc/ANSI_IEEE-Std-754-2019/background/minNum_maxNum_Removal_Demotion_v3.pdf

**Comment by nikic** [comment]
> > So one option would be to upgrade the `f32/f64::min/max` to the 2019 behavior as that's a backwards compatible change (though LLVM doesn't have an intrinsic for it atm).
> 
> it isn't actually backward compatible because IEEE 754-2008 `maxNum` behaves differently than IEEE 754-2019 `maximumNumber` for signalling NaNs: `maxNum` is defined to return a quiet NaN if either input is a signalling NaN, whereas `maximumNumber` returns non-NaN if either input is non-NaN, even if the other input is a signalling NaN.
> 
> background on min/max in IEEE 754-2019: https://grouper.ieee.org/groups/msc/ANSI_IEEE-Std-754-2019/background/minNum_maxNum_Removal_Demotion_v3.pdf

While this is technically true, this comes with the caveat that people don't actually use the IEEE 754-2008 definition of maxNum and instead use "maxNum, but with sNaN treated the same ways as qNaN". So basically, what IEEE 754-2008 maxNum is in practice, is IEEE 754-2019 maximumNumber with unspecified signed zero behavior.

Here is Rust's own definition:

> If one of the arguments is NaN, then the other argument is returned. This follows the IEEE-754 2008 semantics for maxNum, except for handling of signaling NaNs; this function handles all NaNs the same way and avoids maxNum’s problems with associativity. This also matches the behavior of libm’s fmax.

And this is LLVM's more extensive definition:

> Follows the IEEE-754 semantics for minNum, except for handling of signaling NaNs. This match’s the behavior of libm’s fmin.
>
> If either operand is a NaN, returns the other non-NaN operand. Returns NaN only if both operands are NaN. The returned NaN is always quiet. If the operands compare equal, returns a value that compares equal to both operands. This means that fmin(+/-0.0, +/-0.0) could return either -0.0 or 0.0.
>
> Unlike the IEEE-754 2008 behavior, this does not distinguish between signaling and quiet NaN inputs. If a target’s implementation follows the standard and returns a quiet NaN if either input is a signaling NaN, the intrinsic lowering is responsible for quieting the inputs to correctly return the non-NaN input (e.g. by using the equivalent of llvm.canonicalize).

And now we are in the wonderful position where there are actually 5 different notions, all commonly accepted, of what a float maximum is supposed to be :)

**Comment by juliuskoskela** [comment]
I'd like to add to this issue that I ran into `simd_min/simd_max` functions destroying the performance of my algorithm where I had previously used `_mm_min_ps` or the likes. Replacing them with LT/GT comparisons solved the issue, but I feel like the solution was less than obvious.

It's a tricky situation, but at the least the documentations should reflect these performance considerations.

**Comment by scottmcm** [comment]
Note that Intel specifically documents `_mm_min_ps` as

> does not follow the IEEE Standard for Floating-Point Arithmetic (IEEE 754) minimum value when inputs are NaN or signed-zero values.

Whereas [`f32::minimum`](https://doc.rust-lang.org/nightly/std/primitive.f32.html#method.minimum) says it *does* follow the IEEE 754-2019 semantics.

I don't think we should generally add "but it might be faster, on some platforms, to do something that works differently" to method documentation.

**Comment by juliuskoskela** [comment]
> I don't think we should generally add "but it might be faster, on some platforms, to do something that works differently" to method documentation.

I think the correct place to document this could be the beginners guide or if there's gonna be a chapter in "the book" about portable simd.

**Comment by GitSplits** [comment]
Apologies for being so late to the party here, but isn’t this all part of a _behavior_ and therefore better handled as an _attribute_ for the section of code it’s being used within? For example different CPU’s implement different behaviors and also a developer might want to lock-in their code to a certain IEEE 754 standard.

In particular, how often is it that one section of code is going to be interleaving two separate behaviors? Obviously this extends to all the other floating point arithmetic, not just max/min.

Having used other languages where similar words have been used for slightly different meanings has left the code difficult to understand and debug.

Finally, there should be room remaining for a future unexpected kind of number, leading to more and more max/min concepts better covered with and attribute rather than more variants of the words max and min.

**Comment by GitSplits** [comment]
Rather than having alternate names for max and min, why not use attributes instead? This way the attribute would specify how max/min is applied for a certain block of code, with consistent max/min algorithms and a well understood and expected output. This of course assumes different max/min behaviors are seldom intermingled together.

Using attributes also allows for different max/min mechanisms per different CPU types and also for different IEEE 754 standards as they’re published and adjusted over time.

**Comment by eduardosm** [comment]
In https://github.com/rust-lang/rust/pull/130568, I'm making these methods `const`, keeping their constness under the `float_minimum_maximum` feature gate.

**Comment by tgross35** [comment]
Just adding some data points: C23 provides `fminimum_num` and `fmaximum_num`, which correlate with IEEE 754-2018 `minimumNumber` and `maximumNumber`:

> These functions return the number if one argument is a number and the other is a quiet or signaling NaN. If both arguments are NaNs, a quiet NaN is returned. If an argument is a signaling NaN, the "invalid" floating-point exception is raised (even though the function returns the number when the other argument is a number).

It also seems to encourage their use over `fmin` and `fmax`:

> The functions fmin and fmax have been superseded by fminimum_num and fmaximum_num. The fmin and fmax functions provide the minNum and maxNum operations specified in (the superseded) ISO/IEC/IEEE 60559:2011.

`fmaximum` and `fminimum` provide the versions that return NaN if any input is NaN, which is 754-2018 `minimum`/`maximum`.

We could mirror this by keeping what we currently have for names of the new functions (`minimum` and `maximum`) but introducing `maximum_num` and `minimum_num` that provide `minimumNumber` and `minimumNumber`. And then, after a couple of cycles, deprecate `min` and `max` so the names are unambiguous.

It would be unfortunate to lose the short names. We could also do an edition migration for 2027:

- Rename the unstable `minimum` and `maximum` to `min_2027` and `max_2027`
- Add `min_num` and `max_num` that provide the current `min`/`max` semantics. `min`/`max` should call these.
- Create an edition autoupgrade that changes `max` -> `max_num` and `min` -> `min_num`
- At the edition, change `min` and `max` to call `min_2027` and `max_2027`
- Deprecate/hide/whatever the `_2027` versions

... but that may be more trouble than it is worth.

**Comment by programmerjake** [comment]
> It would be unfortunate to lose the short names. We could also do an edition migration for 2027:
> 
> \<lots of complex steps...>
> 
> ... but that may be more trouble than it is worth.

I think a better idea if we want to end up with the short `min`/`max` names is to implement [my suggestion for allowing multiple things with the same name but different editions](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Effect.20of.20fma.20disabled/near/274199236) where you can supply an explicit edition disambiguator to choose which one you want, or just use the default from your edition (it picks the nearest edition, or maybe we could extend the syntax to have edition ranges).

so the standard library would look like (for `min` at least, I'm not saying we can't also add variants with other names):
```rust
impl f64 {
    pub fn e2024#min(self, other: Self) -> Self {
        todo!("equivalent to libc::fmin")
    }
    pub fn e2027#min(self, other: Self) -> Self {
        todo!("equivalent to libc::fminimum")
    }
}
```

**Comment by valadaptive** [comment]
I agree that having three operations is confusing, and we should stick to two. I'd prefer `minimumNumber` and `maximumNumber` for the reasons I gave above, as well as the current `min`/`max` which don't require expensive emulation on x86. Since the specifics of NaN bit patterns don't apply here due to LLVM semantics, the only difference between the operations would be signed zero handling.

(As a side note, it is a bit tragic that the intrinsic operations that the Rust core library exposes, and that *only* the Rust core library can expose because they're intrinsics, are restricted by what makes for the nicest API. It'd be nice to provide architecture-specific scalar intrinsics under `core::arch` or something for the platforms that do provide hardware support.)

**Comment by tgross35** [comment]
To get an idea of what exactly we're looking at optimization-wise, here's a table of how these scalar ops currently get lowered. Based on https://llvm.godbolt.org/z/f3n57WE1q.

| arch      | minNum (2008) f32 | minimum (2019) f32 | minimumNumber (2019) f32 | minNum (2008) f64 | minimum (2019) f64 | minimumNumber (2019) f64 |
|-----------|-------------------|--------------------|--------------------------|-------------------|--------------------|--------------------------|
| x86-64    | routine (short)   | routine            | routine                  | routine (short)   | routine            | routine                  |
| aarch64   | instruction       | instruction        | routine (tiny)           | instruction       | instruction        | routine (tiny)           |
| riscv64gc | instruction       | routine            | instruction              | instruction       | routine            | instruction              |

(Please feel free to edit this message and add more)

**Comment by tgross35** [comment]
> I agree that having three operations is confusing, and we should stick to two. I'd prefer `minimumNumber` and `maximumNumber` for the reasons I gave above, as well as the current `min`/`max` which don't require expensive emulation on x86. Since the specifics of NaN bit patterns don't apply here due to LLVM semantics, the only difference between the operations would be signed zero handling.

Two clarify: you would prefer having two functions that return the non-NAN (YAN?) functions but differ only by their handling of signed zeroes? Current `min` is 2008 `minNum`.

> (As a side note, it is a bit tragic that the intrinsic operations that the Rust core library exposes, and that _only_ the Rust core library can expose because they're intrinsics, are restricted by what makes for the nicest API. It'd be nice to provide architecture-specific scalar intrinsics under `core::arch` or something for the platforms that do provide hardware support.)

I don't think this is the craziest idea. In my ideal world we would have inherent methods `min` (2019 `minimum`) and `min_num` (2019 `minimumNumber`) as since everything will probably move that way, at a molasses pace. I'd feel less bad about removing the 2008 `minNum` semantics if anyone who wants the perf could get it from `core::arch`.

**Comment by nickspiker** [comment]
**The missing intrinsic (#140792) isn't just an ergonomics issue, it gates codegen quality which is a concrete reason to prioritize it and of course the ordering problem where x.min(a).max(b) is not equivalent to x.max(b).min(a) but only when NaN is involved.**

I went down this from a random image pipeline and disassembled the results as it is common to see naieve implementations like the clamp then cast (nightly `1.97.0`, `aarch64-unknown-linux-gnu`, `-O`). The underlying gap is that no mainstream ISA has the IEEE 754-2019 propagating min/max in hardware — they ship the *deprecated* 754-2008 NaN-suppressing op (`fminnm` on aarch64). So:

`f32::min` (suppressing) lowers to one instruction class:
```asm
old_min:
    fminnm s1, s1, s1
    fminnm s0, s0, s0
    fminnm s0, s0, s1
    ret
```

`f32::minimum` (propagating — the correct op) has no single instruction, so it open-codes to a branch tree (~17 insns: `fcmp`/`b.mi`/`fmov`/`cmn`/`ccmp`/`fcsel`/`fadd`):
```asm
new_minimum:
    fcmp s0, s1
    b.mi .LBB0_5
    fcmp s1, s0
    b.pl .LBB0_3
    fmov s0, s1
    ret
    ...
```

The downstream cost: that open-coded blob also defeats the `fptoui.sat`-of-clamp peephole. `x.maximum(0.0).minimum(255.0) as u8` is provably identical to `x as u8` for *every* input (NaN→0 both ways via the saturating cast; >255→255; <0→0), yet LLVM emits ~17 branchy instructions for the clamped form vs 4 for the bare cast. It can't see thru the branch tree to fold the now-dead clamp. With `llvm.minimum`/`llvm.maximum` intrinsics, both the direct op *and* this fold should collapse to the single-instruction path.

So #140792 looks like the real lever: wiring the intrinsics unblocks correctness-availability *and* codegen together. Given the fallback impl is already correct, would it be reasonable to stabilize on the fallback now (correctness win, available on stable) and treat the intrinsic as the follow-up that recovers codegen? Today, reaching for the non-suppressing op requires nightly and is slow even there — a self-reinforcing reason people keep defaulting to the suppressing `min`/`max`.

**Comment by RalfJung** [comment]
You seem to have replied to an outdated part of the tracking issue. I have updated the "unresolved questions". I don't think anyone is concerned about the intrinsic for stabilization.

EDIT: Actually, the intrinsic isn't used as it is buggy. But I agree with you that this should not block stabilization, and I don't think the team considers it a blocker.

**Comment by nickspiker** [comment]
Apologies, I replied to the stale unresolved questions and over focused on the intrinsic. Dropping that.

For me it's strictly a correctness issue. Every other float op propagates NaN, so an undefined result stays observable and you handle it once at the end. `min`/`max` are the lone exception that suppress it, which silently erases "this value is undefined" from a computation. I hit it as silent output corruption in an image pipeline: a NaN from an upstream error got absorbed into a plausible finite value instead of propagating, so the bug stayed invisible.

The other thing that bugs me is that clamp ordering changes the result, but only on NaN. `(NaN).min(255.).max(0.) as u8` gives 255, while `(NaN).max(0.).min(255.) as u8` gives 0. Identical for every finite value, both infinities, both signed zeros, yet opposite extremes (white vs black) for NaN, decided purely by which clamp came first. With propagating `minimum`/`maximum` both give 0, because NaN flows through and the saturating cast lands it at 0. That order independence is what I want, and it's the correct behavior.

So +1 to @ickk: I specifically want the NaN propagating `minimum`/`maximum`.

**Comment by RalfJung** [comment]
Note that we have [`clamp`](https://doc.rust-lang.org/nightly/std/primitive.f32.html#method.clamp) so you shouldn't have to use a min/max sequence. It has the NaN behavior you seem to be looking for.

Propagating or not propagating NaN both can make sense, which is why the latest IEEE standard has both kinds of operations. I agree Rust should also have both. The main remaining blocker seems to be the naming -- but we'd need someone from @rust-lang/libs-api to chime in to be sure.
